x/twap/migrate.go

```diff
@@ -14,3 +32,46 @@ func (k Keeper) MigrateExistingPools(ctx sdk.Context, latestPoolId uint64) error
 	}
 	return nil
 }
+
+// initializeGeometricTwapAcc initializes all existing twap records in state
+// with the given value. Returns nil on success, error otherwise.
+// Returns error if:
+// - fails to get historical time indexed records.
+// - does not find any time indexed records.
+// - the time indexed records returned from state are written back
+// in incorrect order. The order must be ascending by time so that
+// we update the "most recent" index with the right records correctly.
+func (k Keeper) initializeGeometricTwapAcc(ctx sdk.Context, value sdk.Dec) error {
+	// In ascending order by time.
+	historicalTimeIndexed, err := k.getAllHistoricalTimeIndexedTWAPs(ctx)
+	if err != nil {
+		return err
+	}
+
+	if len(historicalTimeIndexed) == 0 {
+		return errors.New("error: no historical twap records found")
+	}
+
+	// Since we are iterate over time-indexed records in ascending order,
+	// most recent record should also be updated correctly.
+	for i, record := range historicalTimeIndexed {
+		record := record
+		// Sanity check order.
+		if i > 0 {
+			previousRecord := historicalTimeIndexed[i-1]
+
+			isInvalidOrder := previousRecord.Time.After(record.Time)
+			if isInvalidOrder {
+				return fmt.Errorf("error: historical twap records are not in ascending order, (%v), was after (%v)", previousRecord, record)
+			}
+		}
+
+		record.GeometricTwapAccumulator = value
```

Comment:
Can/should we check if `record.GeometricTwapAccumulator` already has a value and skip it if it does?

I'm thinking of a future minor version update from state export (like the issue we just had on 13.1.2) that may try to run this and would end up resetting all geometric accums

Reply:
This logic is meant to be run only once. I don't see why and how this would be re-run.

I would be supportive of adding this check. However, it is non-trivial for tests.

The reason is that we are already on v14.x code here. If the proto field exists, then it is initialized to zero.
We cannot set up the test in such a way where we have a v13.x (old) twap record so that its geometric twap accumulator is uninitialized.

Reply:
This makes sense. I think we should make sure that it doesn't get run again on an upgrade from export. I'm not sure if this is what would happen though. Just extrapolating from the issue with `initGenesis` on 13.1.2.